### PR TITLE
fix creating GitLab comments on unmodified lines

### PR DESCRIPTION
**Pull request by alexintech**
This pull request fixes some issues with GitLab Reporter:

- get all existing comments using pagination
- fixing incorrect request `new_line=0,old_line=0` if comment should be created on unmodified line that doesn't exist in the diff

Added extra tests for different cases for creating comments in GitLab.

### CODE REVIEW

File: internal/reporter/gitlab.go [hunk 1]
```diff
@@ -383,10 +392,37 @@ type diffLine struct {
...
-	for _, dl := range lines {
+	if len(lines) == 0 {
+		return diffLine{}, false
+	}
+
+	for i, dl := range lines {
 		if dl.new == line {
 			return dl, true
 		}
+		// Calculate unmodified line that does not present in the diff
```
Comment (prymitive) [inline internal/reporter/gitlab.go:403]: What is going on here? Why are you looking for unmodified lines?

**Comment by prymitive** [comment]
> * get all existing comments using pagination

Can you split this into a dedicated commit so it's easier to review changes in this PR please?

**Comment by alexintech** [comment]
> > * get all existing comments using pagination
> 
> Can you split this into a dedicated commit so it's easier to review changes in this PR please?

sure, done.

File: internal/reporter/gitlab.go (hunk 1 above)
Comment (alexintech) [reply internal/reporter/gitlab.go:403]: Diff commonly includes some lines before changes and some after. If we have small change in a big alert definition, such as:

```
- alert: AlertName
      annotations:
        dashboard: url
        description: some long description
        runbook_url: url
        summary: summary note
      expr: up > 0
      for: 10m
      labels:
        severity: critical
```

For example, if we change only `runbook_url`, diff got from GitLab API may include lines starting `annotations:` and ending `labels:`. 

Pint validates the whole alert and may find problems, that it want's to publish on unmodified lines, that doesn't present in the diff (for example, at the end list of `labels` if some label is required).

In such cases we get an error similar to:
```
level=DEBUG msg="Got pending comment" reporter=GitLab path=alert.yaml line=23 msg=":warning: **Warning** reported by [pint](https://cloudflare.github.io/pint/) **rule/label** check.\n\n------\n\n`alert_for` label is required.\n\n------\n\n:information_source: To see documentation covering this check and instructions on how to resolve it [click here](https://cloudflare.github.io/pint/checks/rule/label.html).\n"
level=DEBUG msg="Comment doesn't exist yet and needs to be created" reporter=GitLab path=alert.yaml line=23
level=DEBUG msg="Creating a new merge request discussion" base_sha=d135e0bfc702c6d51f70394119a29d04c9695063 head_sha=8ddd5b4d620d808c1b748f31c121dd3c6c822517 start_sha=d135e0bfc702c6d51f70394119a29d04c9695063 old_path=alert.yaml new_path=alert.yaml old_line=0 new_line=0
level=DEBUG msg="performing request" reporter=GitLab method=POST url=https://gitlab/api/v4/projects/123/merge_requests/35/discussions
level=ERROR msg="Failed to create a new comment" reporter=GitLab path=alert.yaml line=23 err="POST https://gitlab/api/v4/projects/123/merge_requests/35/discussions: 400 {message: 400 Bad request - Note {:line_code=>[\"can't be blank\", \"must be a valid line code\"]}}"
```

That's because [parseDiffLines()](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aR432) does not have `dl.new`=23, and empty `diffLine{}, false` returns [here](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aL391). In that case [`case !ok`](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aR360) results to incorrect request with `new_line=0 old_line=0`

So the fix [here](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aR404) checks if `line` before the first diff line (or before the next diff block) and calculates `diffLine{}` based on the first diff line information. It just calculates the `gap` [here](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aR409) and adds it to `lastLines.old` to get `old_line` value.

The same goes [here](https://github.com/cloudflare/pint/pull/1147/files#diff-587726f401f87b2143e32dd2b81a98eaca441fdffdb33dca660c8ba27bb0693aR418) for cases if `line` is after the all values in `diffLines`.

Actually, I suppose, that `case !ok` should not be called anymore (and test coverage shows that it is not called). But maybe I don't see some corner case and missed it in [testCases](https://github.com/cloudflare/pint/pull/1147/files#diff-cc0df7059a063851df54350db36e2d7a962af774a8a7d4ecd6bb1b88db275059R439)
